lib/WP_Auth0_UsersRepo.php

```diff
@@ -68,90 +68,83 @@ public function tokenHasRequiredScopes( $jwt ) {
+		$email_verified = ! empty( $userinfo->email_verified )
+			|| $skip_email_verified
+			|| $opts->strategy_skips_verified_email( $strategy );
 
+		// WP user to join with incoming Auth0 user.
 		if ( ! empty( $userinfo->email ) ) {
-			$joinUser = get_user_by( 'email', $userinfo->email );
+			$wp_user = get_user_by( 'email', $userinfo->email );
 		}
 
-		$auto_provisioning = WP_Auth0_Options::Instance()->get( 'auto_provisioning' );
-		$allow_signup      = WP_Auth0_Options::Instance()->is_wp_registration_enabled() || $auto_provisioning;
-
-		$user_id = null;
-
-		global $wpdb;
-
-		// If there is a user with the same email, we should check if the wp user was joined with an auth0 user. If so, we shouldn't allow it again
-		if ( ! is_null( $joinUser ) && $joinUser instanceof WP_User ) {
-			$auth0_id = get_user_meta( $joinUser->ID, $wpdb->prefix . 'auth0_id', true );
-
-			if ( $auth0_id ) { // if it has an a0 id, we cant join it
-				$msg = __( 'There is a user with the same email', 'wp-auth0' );
+		if ( is_object( $wp_user ) && $wp_user instanceof WP_User ) {
+			// WP user exists, check if we can join.
+			$user_id = $wp_user->ID;
 
-				throw new WP_Auth0_CouldNotCreateUserException( $msg );
-			}
-		}
-
-		if ( $shouldJoinUser && ! is_null( $joinUser ) && $joinUser instanceof WP_User ) {
-			// If we are here, we have a potential join user
-			// Don't allow creation or assignation of user if the email is not verified, that would
-			// be hijacking
-			if ( $ignore_unverified_email || $userinfo->email_verified ) {
-				$user_id = $joinUser->ID;
-			} else {
+			// Cannot join a DB connection user without a verified email.
+			if ( $is_db_connection && ! $email_verified ) {
 				throw new WP_Auth0_EmailNotVerifiedException( $userinfo, $token );
 			}
-		} elseif ( $allow_signup ) {
 
-			// If we are here, we need to create the user
+			// If the user has a different Auth0 ID, we cannot join it.
+			$current_auth0_id = self::get_meta( $user_id, 'auth0_id' );
+			if ( ! empty( $current_auth0_id ) && $auth0_sub !== $current_auth0_id ) {
+				throw new WP_Auth0_CouldNotCreateUserException( __( 'There is a user with the same email.', 'wp-auth0' ) );
+			}
+		} elseif ( $opts->is_wp_registration_enabled() || $opts->get( 'auto_provisioning' ) ) {
+			// WP user does not exist and registration is allowed.
 			$user_id = WP_Auth0_Users::create_user( $userinfo, $role );
 
 			// Check if user was created
 			if ( is_wp_error( $user_id ) ) {
 				throw new WP_Auth0_CouldNotCreateUserException( $user_id->get_error_message() );
-			} elseif ( $user_id == -2 ) {
-				$msg = __( 'Could not create user. The registration process were rejected. Please verify that your account is whitelisted for this system. Please contact your site’s administrator.', 'wp-auth0' );
-
-				throw new WP_Auth0_CouldNotCreateUserException( $msg );
+			} elseif ( -2 === $user_id ) {
```

Comment:
Where does -2 come from? Could this be a constant that better defines what this means?

Reply:
It's what `WP_Auth0_Users::create_user()` returns for that condition. I didn't want to go much further with this PR but I totally agree that this magic value is not ideal 👍